### ISSUE TITLE
exhaustive checking?

### ISSUE THREAD

**Issue by unhammer** [carymrobbins/haskell-conf#3]
Is there a way to make it fail loading if some value is not exhaustive?

E.g. say I want to require that a config has a string for every alternative of `data D = A | B | C`. If I just do  `strings=[(A,"astring"),(B,"bstring")]`, it'll load fine, but at some point during runtime I'll get a failed lookup for C – it'd be nice to see that on startup instead.

**Comment by carymrobbins** [comment]
It's probably easiest to just define this with a helper function.  You could add `Bounded, Enum, Eq, Ord` to your `data D` and you can then generate a set of all possible values -

``` haskell
import Data.Set (Set)
import qualified Data.Set as S

data D = A | B | C deriving (Read, Show, Bounded, Enum, Eq, Ord)

allDs :: Set D
allDs = S.fromList [minBound..maxBound]
```

You could then just define a function for finding the missing keys (or however you want to validate it) -

``` haskell
missingDs :: [(D, a)] -> Set D
missingDs xs = S.difference allDs $ S.fromList $ fst <$> xs
```

Example -

``` haskell
exampleValue =
  [ (A, "foo")
  , (B, "bar")
  ]

main =
  if S.null $ missingDs exampleValue then
    putStrLn "OK"
  else
    putStrLn $ "Missing D keys: " ++ show (missingDs exampleValue)
```

If your key data type is a bit more complex, you might want to look at the [Data.Data](https://hackage.haskell.org/package/base-4.8.2.0/docs/Data-Data.html) module for some generalized constructor inspection using `Typeable`.

Long story short, I don't think this should live as part of this package as this is a very specialized use case.  For many data types, this just wouldn't be possible as exhaustiveness checking doesn't occur in places like this.

**Comment by unhammer** [comment]
Aha, thanks a bunch for the detailed answer :-)

**Comment by carymrobbins** [comment]
No problem!
